crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -393,10 +549,146 @@ impl Drop for MemoryPool {
+    // `PROT_NONE` region for it before any memory slots. Recall that we can
+    // avoid bounds checks for loads and stores with immediates up to
+    // `guard_bytes`, but we rely on Wasmtime to emit bounds checks for any
+    // accesses greater than this.
+    let pre_slab_guard_bytes = if guard_before_slots { guard_bytes } else { 0 };
+
+    let (num_stripes, needed_guard_bytes) = if guard_bytes == 0 || max_memory_bytes == 0 {
+        // In the uncommon case where the memory or guard regions are empty, we
+        // will not need any stripes: we just lay out the slots back-to-back
+        // using a single stripe.
+        (1, guard_bytes)
+    } else if num_pkeys_available < 2 {
+        // If we do not have enough protection keys to stripe the memory, we do
+        // the same. We can't elide any of the guard bytes because we aren't
+        // overlapping guard regions with other stripes...
+        (1, guard_bytes)
+    } else if num_memory_slots == 0 {
+        (1, guard_bytes)
+    } else {
+        // ...but if we can create at least two stripes, we can use another
+        // stripe (i.e., a different pkey) as this slot's guard region--this
+        // reduces the guard bytes each slot has to allocate. We must make sure,
+        // though, that if the size of that other stripe(s) does not fully cover
+        // `guard_bytes`, we keep those around to prevent OOB access.
+        //
+        // We first calculate the number of stripes we need: we want to minimize
+        // this so that there is less chance of a single store running out of
+        // slots with its stripe--we need at least two, though. But this is not
+        // just an optimization; we need to handle the case when there are fewer
+        // slots than stripes. E.g., if our pool is configured with only three
+        // slots (`num_memory_slots = 3`), we will run into failures if we
+        // attempt to set up more than three stripes.
+        let needed_num_stripes =
+            guard_bytes / max_memory_bytes + usize::from(guard_bytes % max_memory_bytes != 0) + 1;
+        let num_stripes = num_pkeys_available
+            .min(needed_num_stripes)
+            .min(num_memory_slots);
+        let next_slots_overlapping_bytes = max_memory_bytes
+            .checked_mul(num_stripes - 1)
+            .unwrap_or(usize::MAX);
+        let needed_guard_bytes = guard_bytes
+            .checked_sub(next_slots_overlapping_bytes)
+            .unwrap_or(0);
+        (num_stripes, needed_guard_bytes)
+    };
+
+    // The page-aligned slot size; equivalent to `memory_and_guard_size`.
+    let page_alignment = crate::page_size() - 1;
+    let slot_bytes = max_memory_bytes
+        .checked_add(needed_guard_bytes)
+        .and_then(|slot_bytes| slot_bytes.checked_add(page_alignment))
+        .and_then(|slot_bytes| Some(slot_bytes & !page_alignment))
+        .ok_or_else(|| anyhow!("slot size is too large"))?;
```

Comment:
I'd be a bit hesitant about doing some of the above calculations and then page-aligning afterwards. Could the inputs all be page-aligned at the beginning as appropriate instead?

Reply:
`max_memory_bytes` should be Wasm-page aligned but we need to realign after all the division, subtraction, etc.